avr-hal-generic/src/adc.rs

```diff
@@ -1,11 +1,35 @@
-/// Analog-to-Digial converter
+//! Analog-to-Digial Converter (ADC)
+//!
+//! # Basic information
+//!
+//! The AVR chips have ADCs, which allow the CPU to acquire information about
+//! the "intensity" of a signal, in this case via voltage measurement. This is
+//! in contrast to a digital input, which only acquires information about
+//! whether there is signal or not.
+//!
+//! To do this, the converter has circuitry to transform this single continuous
+//! signal into multiple discrete signals which the CPU can understand. These
+//! signals map to increasing digits of a binary integer, and interpolating this
+//! up-counter to a known scale yields "analog" information the chip can use.
+//!
+//! # Advanced information
+//!
+//! Due to size and resource constraints, some complexities are introduced to
+//! aliviate issues in manufacturing. In this case, there's only one CPU
+//! register to read all conversions, and the various ADC channels are
+//! multiplexed to it via selection registers. Also, each channel can read from
+//! any of a few pin choices, again via multiplexing with a selector.
+//! Furthermore, we can choose to measure voltage with respect to references
+//! other than the system GND. This module accounts for all these choices
+//! statically with optional dynamic casting.
```

Comment:
Very kind of you to document all that but for the future, you don't need to go to such length :)  If people don't know how an ADC works, they need to learn a lot more things anyway so I don't think it is our job to explain this here.

In `arduino-hal` a more detailed explanation may be nice to have as that's user-facing and many beginner embedded devs will try using it.  But when someone digs into `avr-hal-generic`, they will need to have more embedded background anyway ;)

Reply:
I think we can move this to arduino-hal, that does seem like a better place.